lib/srv/db/mongodb/autousers_admin.go

```diff
@@ -0,0 +1,260 @@
+	databaseName string
+	clock        clockwork.Clock
+	log          logrus.FieldLogger
+	cleanupTTL   time.Duration
+}
+
+// checkAndSetDefaults validates config and sets defaults.
+func (c *shareableAdminClientConfig) checkAndSetDefaults() error {
+	if c.rawClient == nil {
+		return trace.BadParameter("missing mongo.Client")
+	}
+	if c.adminUser == "" {
+		return trace.BadParameter("missing adminUser")
+	}
+	if c.databaseName == "" {
+		return trace.BadParameter("missing databaseName")
+	}
+	if c.clock == nil {
+		c.clock = clockwork.NewRealClock()
+	}
+	if c.log == nil {
+		c.log = logrus.StandardLogger()
+	}
+	if c.cleanupTTL <= 0 {
+		c.cleanupTTL = adminClientCleanupTTL
+	}
+	return nil
+}
+
+type shareableAdminClient struct {
+	adminClient
+	shareableAdminClientConfig
+	wg    sync.WaitGroup
+	timer clockwork.Timer
+}
+
+func newShareableAdminClient(cfg shareableAdminClientConfig) (*shareableAdminClient, error) {
+	if err := cfg.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	c := &shareableAdminClient{
+		adminClient:                cfg.rawClient,
+		shareableAdminClientConfig: cfg,
+		timer:                      cfg.clock.NewTimer(cfg.cleanupTTL),
+	}
+
+	go c.waitAndCleanup()
+	return c, nil
+}
+
+func (c *shareableAdminClient) waitAndCleanup() {
+	c.log.Debugf("Created new MongoDB connection as admin user %q on %q.", c.adminUser, c.databaseName)
+
+	// Wait until TTL.
+	<-c.timer.Chan()
+
+	// Wait until all shared sessions are released.
+	c.wg.Wait()
+
+	// Disconnect.
+	if err := c.adminClient.Disconnect(context.Background()); err != nil {
+		c.log.Warnf("Failed to disconnect MongoDB connection as admin user %q on %q: %v.", c.adminUser, c.databaseName, err)
+	} else {
+		c.log.Debugf("Terminated a MongoDB connection as admin user %q on %q.", c.adminUser, c.databaseName)
+	}
+}
+
+func (c *shareableAdminClient) acquire() *shareableAdminClient {
+	c.wg.Add(1)
+	return c
+}
+
+func (c *shareableAdminClient) Disconnect(_ context.Context) error {
+	c.wg.Done()
+	return nil
+}
+
+type makeBasicAdminClientFunc func(context.Context, *common.Session, *Engine) (adminClient, error)
+
+func makeBasicAdminClient(ctx context.Context, sessionCtx *common.Session, e *Engine) (adminClient, error) {
+	sessionCtx = sessionCtx.WithUser(sessionCtx.Database.GetAdminUser().Name)
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clientCfg, err := makeClientOptionsFromDatabaseURI(sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clientCfg.SetTLSConfig(tlsConfig)
+	clientCfg.SetAuth(mongooptions.Credential{
+		AuthMechanism: mongoauth.MongoDBX509,
+		Username:      x509Username(sessionCtx),
+	})
+
+	client, err := mongo.Connect(ctx, clientCfg)
+	return newBasicAdminClient(client), trace.Wrap(err)
+}
+
+// basicAdminClient is a simple wrapper of mongo.Client with some helper
+// functions.
+type basicAdminClient struct {
+	*mongo.Client
```

Comment:
At first I was afraid that mongo.Client is not thread safe but found this comment in mongo library:

```
// Client is a handle representing a pool of connections to a MongoDB deployment. It is safe for concurrent use by
// multiple goroutines.

type Client struct {
```

Reply:
Yes. Let me add comment on this.